### ISSUE TITLE
Feature request for cif2mtz

### ISSUE THREAD

**Issue by CV-GPhL** [project-gemmi/gemmi#261]
There are quite a few deposited reflection data CIF file at the PDB that have more than an asymmetric unit of reflection data in the _refln category. Although that category is maybe not clearly defined as containing purely merged data (i.e. symmetry-unique reflections for the given spacegroup), it might be useful for "gemmi cif2mtz" to provide some warning messages when running "gemmi cif2mtz"?

Some examples (ignoring cases where systematic absent reflections - e.g. screw axes - are present):
```
 PDB      Nuniq   Ndeposit   Ratio
--------------------------------------
2v96      54936     104532    1.90
3c7r      16457      31550    1.92
3r8u      19429     134339    6.91
3r9h      18458     136229    7.38
3rk9      24357     136479    5.60
3rmf      28852     198931    6.89
3rpv      26361     131060    4.97
3rpy      22474     157554    7.01
3s1h      29112     201426    6.92
4cmu      36299      68369    1.88
4hif      19822      36511    1.84
4hw3      90126     133172    1.48
4i2p      56614      87088    1.54
4i3c      73362      88552    1.21
4jiy       2629       4813    1.83
4m8e      10053      18269    1.82
4m8h      12003      21192    1.77
4nqg      81992     106782    1.30
4poj      17306      31765    1.84
4qrc      24526      37168    1.52
4rfd      30488      46386    1.52
4xs9      49811      95026    1.91
4z3k     148809    1656946   11.13
5k0x      30984      45098    1.46
5kky      38514      58562    1.52
5kz0      16278      29170    1.79
5liu      36038      62556    1.74
6ayh      11504      21003    1.83
6m7h      26747      44916    1.68
6slv      32159      53104    1.65
6veo       8317      14019    1.69
6vka      23873      47248    1.98
6xvw      51860      97755    1.88
7b1l      40624      75440    1.86
7p3v      25348      47514    1.87
7r7f      62933     121130    1.92
7r7g      54713     105503    1.93
7zz9      22752      42666    1.88
7zzb      40162      75210    1.87
7zzc      16965      31865    1.88
7zzf      11351      20191    1.78
7zzg      13027      23356    1.79
```
A ratio of roughly 2 could be due to F+/F- values being stored in "incorrect" data items - but a factor of >7 (3r9h) with deposited amplitudes seems rather odd: truly unmerged data should ideally be in a diffrn_refln loop, but even if placed into the refln loop it seems very peculiar to provide unmerged amplitudes (since merging nearly always happens at the level of intensities and those merged intensities then get converted into amplitudes).

**Comment by wojdyr** [comment]
In the past _refln used to be used for both anomalous and unmerged data, but normally it wasn't in the first data block.
Actually I was asked before to automatically convert reflection cif files such as 1gyu to MTZ file with both F (from the first block) AND F+/F- (from the second block). But I never got to do it.

As for the warning, we have a function that checks if there are non-unique reflections in _refln. It's used with option --list:

    gemmi cif2mtz --list $PDB_DIR/structure_factors/gy/r1gyusf.ent.gz

and prints, for example:

      NOTE: this is old-style anomalous data.

This function checks if all reflections are unique. I could call it when converting files, but I'd need to check if it doesn't slow down significantly the conversion. Probably it doesn't. I suppose a faster method would be possible that would *approximately* check for uniqueness. But then it may not be worth to try to optimize it.

Actually, in case of anomalous data, instead of the warning the data could be automatically converted to F+/F-.
On the other hand, it would be simpler if the PDB remediated all such data files.

**Comment by wojdyr** [comment]
I added a warning.

**Comment by CV-GPhL** [comment]
Hi Marcin,

sorry for the late reply ...

On Thu, Apr 13, 2023 at 09:57:57AM -0700, Marcin Wojdyr wrote:
> In the past _refln used to be used for both anomalous and unmerged
> data, but normally it wasn't in the first data block.

Do you have a record (or even only: rough idea) how many entries
(mis)used the _refln block in that way, i.e. used the incorrect item
name? Or if there is/was a definite date to distinguish "the past"
from current usage?

> Actually I was asked before to automatically convert reflection cif
> files such as 1gyu to MTZ file with both F (from the first block)
> AND F+/F- (from the second block). But I never got to do it.

1gyu also shows two different cell dimensions for the two data blocks
- so the second block is probably not directly related to the first
one (maybe a derivative, since solved by MIR). The second one should
have pdbx_F_minus/pdbx_F_minus_sigma and pdbx_F_plus/pdbx_F_plus_sigma
for unique HKLs (instead of F_meas_au for HKL and -H-K-L).

> As for the warning, we have a function that checks if there are non-unique reflections in _refln. It's used with option --list:
> 
>     gemmi cif2mtz --list $PDB_DIR/structure_factors/gy/r1gyusf.ent.gz
> 
> and prints, for example:
> 
>       NOTE: this is old-style anomalous data.

Ok. Maybe not the clearest text (what does "old-style anomalous data"
really mean) and you could be a bit more verbose with something like

  NOTE: this seems to be old-style anomalous data with refelctions
        (h,k,l) and (-h,k,l) given F_meas_au/F_Meas_sigma_au instead
        of (h,k,l) given pdbx_F_plus/pdbx_F_plus_sigma and
        pdbx_F_minus/pdbx_F_minus_sigma

or (shorter)

  NOTE: this seems to be old-style anomalous data with separate
        reflections (h,k,l) and (-h,k,l) given (F/SIGF) instead of a
        single (h,k,l) with F+/SIGF+ and F-/SIGF-
                          
or similar?

> This function checks if all reflections are unique. I could call it
> when converting files, but I'd need to check if it doesn't slow down
> significantly the conversion.

I'm a big fan of calling as many checking functions as possible when
it comes to mmCIF->MTZ conversion: without warning messages being
thrown back to the user the quality of those reflection mmCIF files
will only improvely slowly I think.

> Probably it doesn't. I suppose a faster method would be possible
> that would *approximately* check for uniqueness. But then it may not
> be worth to try to optimize it.

A check that would also tell a user if the data still contains
systematic absences would be useful too.

> Actually, in case of anomalous data, instead of the warning the data could be automatically converted to F+/F-.

It could (plus DANO/SIGDANO,
i.e. pdbx_anom_difference/pdbx_anom_difference_sigma) - good idea.

> On the other hand, it would be simpler if the PDB remediated all
> such data files.

Absolutely agree: an extensive set of checks and additional
capabilities to the gemmi cif2mtz feature would (1) highlight the need
for remediation and (2) provide the capability to do the remediation
itself.

Of course, once all mmCIF files are always correct/perfect those
checks and features will become obsolete ... but I wouldn't bank on
such a blissful state coming to life in the next years ;-)

Cheers

Clemens

**Comment by wojdyr** [comment]
> Do you have a record (or even only: rough idea) how many entries (mis)used the _refln block in that way, i.e. used the incorrect item name?

I'd need to write a script and run it on the PDB archive. Which I may do later.

>Or if there is/was a definite date to distinguish "the past" from current usage?

No idea.

> 1gyu also shows two different cell dimensions for the two data blocks - so the second block is probably not directly related to the first one (maybe a derivative, since solved by MIR).

Good to know. I didn't realize it before. A couple years ago I got a suggestion that gemmi cif2mtz should do the same as iotbx.cif_as_mtz (phenix.cif_as_mtz). If I run:

    $ iotbx.cif_as_mtz r1gyusf.ent.gz

I get r1gyusf.mtz with the following labels:
 H	K	L	R-free-flags	FOBS	SIGFOBS	F(+)	SIGF(+)	F(-)	SIGF(-)
where FOBS is taken from the first block, and F+/- is from the second block.
I assumed it's a good idea, but apparently it's not.

**Comment by CV-GPhL** [comment]
Hi Marcin,

On Tue, Apr 18, 2023 at 03:49:47AM -0700, Marcin Wojdyr wrote:
> > 1gyu also shows two different cell dimensions for the two data blocks - so the second block is probably not directly related to the first one (maybe a derivative, since solved by MIR).
> 
> Good to know. I didn't realize it before. A couple years ago I got a suggestion that gemmi cif2mtz should do the same as iotbx.cif_as_mtz (phenix.cif_as_mtz). If I run:
> 
>     $ iotbx.cif_as_mtz r1gyusf.ent.gz
> 
> I get r1gyusf.mtz with the following labels:
>  H	K	L	R-free-flags	FOBS	SIGFOBS	F(+)	SIGF(+)	F(-)	SIGF(-)
> where FOBS is taken from the first block, and F+/- is from the second block.

I would not (as a default) mix datablocks - unless there is a clear
description how they are related 9and most of the time there isn't).

> I assumed it's a good idea, but apparently it's not.

It could be - if such a mmCIF file was generated with the same
software that then tries to convert it back (i.e. a round-robin
conversion).

What might be a very good idea is to convert the whole set of SF mmCIF
files with all available (and heavily used) tools out there: "gemmi
cif2mtz", iotbx.cif_as_mtz, our "fetch_PDB" (and we have a rewrite
that is based more on gemmi than CCP4 cif2mtz), maybe PDB-REDO has
something else. There are so many caveats and interpretations
avaialbel, that probably no tool is getting it right in all cases. It
could form a good basis for establishing a standard-operating
procedure for CIF->MTZ conversion - and most likely also provide
insights into the backwards conversion.

Cheers

Clemens

**Comment by wojdyr** [comment]
> Do you have a record (or even only: rough idea) how many entries (mis)used the _refln block in that way, i.e. used the incorrect item name?

2071 blocks store Friedel pairs (presumably anomalous data),
4887 blocks store more multirecord data (presumably unmerged data).

In comparison, I counted only 832 blocks with _diffrn_refln.

For the record, I used current PDB archive, skipping:
- 118 entries that are missing space group name or (in a few cases) missing Miller indices or have incorrect syntax. 
- 6 entries that have files >3GB. They happen to be PanDDA ground-state models – I think with only merged datasets.
